How many late days are you using for this assignment? 
0

In [ ]:
#Repo ID: https://github.com/<redacted>/assignment-3-data-cleaning-and-preprocessing-<redacted>
#Commit ID: 6b63343

In [4]:
# Imports and pip installations (if needed)
%pip install pandas
import pandas as pd

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.


# Part 1: Load the dataset

In [6]:
# Load the given datasets

dataCat = pd.read_csv('chronic_kidney_disease_categorical.csv')
dataNum = pd.read_csv('chronic_kidney_disease_numerical.csv')

# Print the data

print(dataCat)
print()
print(dataNum)

     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
425     769512  0.0  0.0    normal    normal  notpresent  notpresent   no   
426     872034  2.0  1.0  abnormal  abnormal  notpresent     present  yes   
427     643007  NaN  NaN       NaN    normal  notpresent  notpresent  yes   
428     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
429     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm  cad appet   pe  ane  Target  
0    yes   no  good  yes   no     c

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: The datasets represent medical data collected from 400 patients related to chronic kidney disease. The first dataset, chronic_kidney_disease_categorical.csv, contains only categorical features, which have either categorical values or binary values.This dataset records the presence/group of the label. Some of these categorical features are rbc (red blood cells, either normal or abnormal), ba (bacteria, either present or notpresent), and ane (anemia, either yes or no). Furthermore, this dataset contains the target diagnosis for the associated unique_id, either ckd or notckd. The second dataset, chronic_kidney_disease_numerical.csv, contains only numerical features, which have either integer values or continuous values. This dataset records the numerical value of the lavel. Some of these numerical features are age, bp (blood pressure), and rbcc (red blood cell count).

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [7]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows

duplicatesNum = dataNum.duplicated(subset=['unique_id'])
count = duplicatesNum.sum()
print(count)

# Drop these duplicate rows

dataNum = dataNum.drop_duplicates(subset=['unique_id'])

# Repeat the same for categorical dataset. Print the duplicate rows and drop them

duplicatesCat = dataCat.duplicated(subset=['unique_id'])
count = duplicatesCat.sum()
print(count)

dataCat = dataCat.drop_duplicates(subset=['unique_id'])

30
30


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: It's possible for not all unique ids to be present in both datasets, as the dataset information specifies that every variable can have missing values other than the target. If no numerical data is recorded for a given unique_id, then that unique_id will be in the categorical dataset but not the numerical one. In the combined table, the rows that are missing from one of the datasets appear, they just have no data recorded for those labels. For example, the row for a unique_id not in the numerical dataset will just have all the numerical labels have no values.

In [8]:
# Merge the two given numerical and categorical datasets based on their unique_ID.

data = pd.merge(dataCat, dataNum, how='outer', on=['unique_id'])

#Print the combined dataset

print(data)

     unique_id   al   su     rbc        pc         pcc          ba  htn   dm  \
0       102677  3.0  0.0  normal  abnormal  notpresent  notpresent  yes   no   
1       103570  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
2       104675  2.0  1.0  normal    normal     present  notpresent   no  yes   
3       105923  1.0  0.0     NaN    normal  notpresent  notpresent  yes  yes   
4       107297  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
..         ...  ...  ...     ...       ...         ...         ...  ...  ...   
395     995177  4.0  3.0  normal  abnormal     present     present  yes  yes   
396     996076  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
397     996412  NaN  NaN     NaN       NaN  notpresent  notpresent   no   no   
398     997629  0.0  0.0  normal    normal         NaN         NaN   no   no   
399     997876  0.0  0.0     NaN    normal  notpresent  notpresent   no   no   

     cad  ...    bp    bgr     bu    sc

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [9]:
# Calculate the percentage of rows that contain atleast one missing value

rowMissing = data.isnull().any(axis=1)
percentMissing = rowMissing.sum()/len(data) * 100

# Print %

print(percentMissing)

# Drop these rows from the dataset

data = data.dropna()

# Print the Dataset

print(data)

61.75000000000001
     unique_id   al   su     rbc        pc         pcc          ba  htn   dm  \
0       102677  3.0  0.0  normal  abnormal  notpresent  notpresent  yes   no   
6       109053  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
7       114717  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
8       118191  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
9       118596  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
..         ...  ...  ...     ...       ...         ...         ...  ...  ...   
390     980291  3.0  0.0  normal  abnormal  notpresent  notpresent  yes  yes   
391     991031  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
394     994377  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
395     995177  4.0  3.0  normal  abnormal     present     present  yes  yes   
396     996076  0.0  0.0  normal    normal  notpresent  notpresent   no   no   

     cad  ...    bp  

## Part 3.4: Sort the dataset according to the Labels

In [10]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting

data = data.sort_values(by=['Target'])
data = data.reset_index(drop=True)

# Print the dataset

print(data)

     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       102677  3.0  0.0    normal  abnormal  notpresent  notpresent  yes   
1       337834  4.0  0.0    normal  abnormal     present  notpresent  yes   
2       343710  2.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       349892  1.0  3.0  abnormal  abnormal  notpresent  notpresent  yes   
4       353025  3.0  4.0    normal    normal  notpresent  notpresent  yes   
..         ...  ...  ...       ...       ...         ...         ...  ...   
148     486397  0.0  0.0    normal    normal  notpresent  notpresent   no   
149     203466  0.0  0.0    normal    normal  notpresent  notpresent   no   
150     482105  0.0  0.0    normal    normal  notpresent  notpresent   no   
151     228849  0.0  0.0    normal    normal  notpresent  notpresent   no   
152     996076  0.0  0.0    normal    normal  notpresent  notpresent   no   

      dm  cad  ...    bp    bgr     bu    sc    sod  pot  hemo   pcv     wb

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [11]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']

dummy_data = pd.get_dummies(data[categorical_columns], prefix=categorical_columns, prefix_sep='_', dtype=int)
data = pd.concat([data, dummy_data], axis=1)
data = data.drop(categorical_columns, axis=1)

# Print the dataset

print(data)

     unique_id   al   su   age    bp    bgr     bu    sc    sod  pot  ...  \
0       102677  3.0  0.0  59.0  70.0   76.0  186.0  15.0  135.0  7.6  ...   
1       337834  4.0  0.0  48.0  70.0  117.0   56.0   3.8  111.0  2.5  ...   
2       343710  2.0  0.0  61.0  80.0  173.0  148.0   3.9  135.0  5.2  ...   
3       349892  1.0  3.0  59.0  70.0  424.0   55.0   1.7  138.0  4.5  ...   
4       353025  3.0  4.0  40.0  70.0  253.0  150.0  11.9  132.0  5.6  ...   
..         ...  ...  ...   ...   ...    ...    ...   ...    ...  ...  ...   
148     486397  0.0  0.0  20.0  70.0  123.0   44.0   1.0  135.0  3.8  ...   
149     203466  0.0  0.0  22.0  60.0   97.0   18.0   1.2  138.0  4.3  ...   
150     482105  0.0  0.0  44.0  60.0   95.0   46.0   0.5  138.0  4.2  ...   
151     228849  0.0  0.0  58.0  80.0  100.0   50.0   1.2  140.0  3.5  ...   
152     996076  0.0  0.0  57.0  60.0  105.0   49.0   1.2  150.0  4.7  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no  pe_yes  ane_no  ane_ye

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: For a binary variable, this would be equivalent, as one column is able to sufficiently represent the relationship between the possible states of the variable (it is either one thing, or the other). However, if the variable could have more than 2 values, this approach would not work. This is because assigning numerical values to the categorical features in one column could incorrently lead the model to interpreting an ordinal relationship between the categories, with condiment 3 being "better than condiment 2 is. Essentially, this approach could introduce false relationships between categories.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [12]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

for column in numerical_columns:
    mean = data[column].mean()
    std = data[column].std()
    
    upper = mean + 3 * std
    lower = mean - 3 * std
    
    data = data[(data[column] >= lower) & (data[column] <= upper)]

data = data.reset_index(drop=True)

# Print the dataset

print(data)

     unique_id   al   su   age     bp    bgr     bu   sc    sod  pot  ...  \
0       343710  2.0  0.0  61.0   80.0  173.0  148.0  3.9  135.0  5.2  ...   
1       397388  2.0  2.0  63.0  100.0  280.0   35.0  3.2  143.0  3.5  ...   
2       438182  2.0  0.0  73.0   80.0  253.0  142.0  4.6  138.0  5.8  ...   
3       474407  4.0  0.0  21.0   90.0  107.0   40.0  1.7  125.0  3.5  ...   
4       475200  4.0  2.0  64.0  100.0  163.0   54.0  7.2  140.0  4.6  ...   
..         ...  ...  ...   ...    ...    ...    ...  ...    ...  ...  ...   
128     486397  0.0  0.0  20.0   70.0  123.0   44.0  1.0  135.0  3.8  ...   
129     203466  0.0  0.0  22.0   60.0   97.0   18.0  1.2  138.0  4.3  ...   
130     482105  0.0  0.0  44.0   60.0   95.0   46.0  0.5  138.0  4.2  ...   
131     228849  0.0  0.0  58.0   80.0  100.0   50.0  1.2  140.0  3.5  ...   
132     996076  0.0  0.0  57.0   60.0  105.0   49.0  1.2  150.0  4.7  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no  pe_yes  ane_no  ane_ye

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [13]:
# Normalize the all Numerical Attributes in the dataset.
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

for column in numerical_columns:
    data[column] = (data[column] - data[column].min()) / (data[column].max() - data[column].min())

# Print the dataset

print(data)

     unique_id   al   su       age    bp       bgr        bu        sc  \
0       343710  2.0  0.0  0.743243  0.50  0.442060  0.901961  0.432099   
1       397388  2.0  2.0  0.770270  1.00  0.901288  0.163399  0.345679   
2       438182  2.0  0.0  0.905405  0.50  0.785408  0.862745  0.518519   
3       474407  4.0  0.0  0.202703  0.75  0.158798  0.196078  0.160494   
4       475200  4.0  2.0  0.783784  1.00  0.399142  0.287582  0.839506   
..         ...  ...  ...       ...   ...       ...       ...       ...   
128     486397  0.0  0.0  0.189189  0.25  0.227468  0.222222  0.074074   
129     203466  0.0  0.0  0.216216  0.00  0.115880  0.052288  0.098765   
130     482105  0.0  0.0  0.513514  0.00  0.107296  0.235294  0.012346   
131     228849  0.0  0.0  0.702703  0.50  0.128755  0.261438  0.098765   
132     996076  0.0  0.0  0.689189  0.00  0.150215  0.254902  0.098765   

          sod       pot  ...  cad_no  cad_yes  appet_good  appet_poor  pe_no  \
0    0.500000  0.793103  ...   

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: The unique_id column should be removed from the dataset. This column is used as an identifier of a specific patient's data who's value has no relation to the presence of chronic kidney disease or not. Therefore, it should be removed to prevent the predictions from being affected. 

In [14]:
#Remove that column

data = data.drop(columns=['unique_id'])

# Print the dataset

print(data)

      al   su       age    bp       bgr        bu        sc       sod  \
0    2.0  0.0  0.743243  0.50  0.442060  0.901961  0.432099  0.500000   
1    2.0  2.0  0.770270  1.00  0.901288  0.163399  0.345679  0.766667   
2    2.0  0.0  0.905405  0.50  0.785408  0.862745  0.518519  0.600000   
3    4.0  0.0  0.202703  0.75  0.158798  0.196078  0.160494  0.166667   
4    4.0  2.0  0.783784  1.00  0.399142  0.287582  0.839506  0.666667   
..   ...  ...       ...   ...       ...       ...       ...       ...   
128  0.0  0.0  0.189189  0.25  0.227468  0.222222  0.074074  0.500000   
129  0.0  0.0  0.216216  0.00  0.115880  0.052288  0.098765  0.600000   
130  0.0  0.0  0.513514  0.00  0.107296  0.235294  0.012346  0.600000   
131  0.0  0.0  0.702703  0.50  0.128755  0.261438  0.098765  0.666667   
132  0.0  0.0  0.689189  0.00  0.150215  0.254902  0.098765  1.000000   

          pot      hemo  ...  cad_no  cad_yes  appet_good  appet_poor  pe_no  \
0    0.793103  0.000000  ...       0       

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [15]:
# Export the dataframe to a new csv file

data.to_csv('chronic_kidney_disease_clean.csv', index=False)

# Export the dataframe to a new json file

data.to_json('chronic_kidney_disease_clean.json', orient="records")


# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

Paste here: 

### Paste the markdown table here

| al | su | age | bp | bgr | bu | sc | sod | pot | hemo | pcv | wbcc | rbcc | rbc_abnormal | rbc_normal | pc_abnormal | pc_normal | pcc_notpresent | pcc_present | ba_notpresent | ba_present | htn_no | htn_yes | dm_no | dm_yes | cad_no | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd | Target_notckd |
|-----|-----|------------|------|--------------------|-----------------------|---------------------|---------------------|---------------------|---------------------|---------------------|---------------------|---------------------|--------------|------------|--------------|------------|-----------------|--------------|-----------------|------------|--------|---------|-------|--------|--------|---------|------------|------------|-------|--------|--------|---------|-------------|----------------|
| 3.0 | 0.0 | 0.87837838 | 0.25 | 0.6394849785407726 | 0.47058823529411764 | 0.39506172839506176 | 0.43333333333333335 | 0.517241379310345 | 0.2673267326732673 | 0.3225806451612903 | 0.10483870967741936 | 0.17142857142857146 | 0 | 1 | 1 | 0 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 1 | 0 | 1 | 0 | 1 | 0 |
| 1.0 | 0.0 | 0.71621622 | 0.50 | 1.0 | 0.16339869281045752 | 0.11111111111111112 | 0.06666666666666667 | 0.20689655172413796 | 0.2673267326732673 | 0.3870967741935484 | 0.532258064516129 | 0.3714285714285714 | 1 | 0 | 0 | 1 | 1 | 0 | 1 | 0 | 1 | 0 | 0 | 1 | 1 | 0 | 0 | 1 | 1 | 0 | 1 | 0 | 1 | 0 |
| 4.0 | 0.0 | 0.00 | 0.00 | 0.10300429184549356| 0.37254901960784315 | 0.07407407407407407 | 0.5 | 0.6896551724137933 | 0.21782178217821782 | 0.22580645161290322 | 1.0 | 0.5142857142857142 | 1 | 0 | 1 | 0 | 1 | 0 | 0 | 1 | 1 | 0 | 1 | 0 | 1 | 0 | 0 | 1 | 1 | 0 | 1 | 0 | 1 | 0 |
| 4.0 | 0.0 | 0.87837838 | 0.0 | 0.20600858369098712| 0.7516339869281046 | 0.6049382716049382 | 0.5333333333333333 | 0.6896551724137933 | 0.36633663366336633 | 0.3870967741935484 | 0.8790322580645161 | 0.3714285714285714 | 0 | 1 | 0 | 1 | 1 | 0 | 1 | 0 | 0 | 1 | 0 | 1 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 |
| 2.0 | 0.0 | 0.72972973 | 0.75 | 0.15021459227467812| 0.28104575163398693 | 0.2345679012345679 | 0.5333333333333333 | 0.7931034482758622 | 0.33663366336633654 | 0.3225806451612903 | 0.5 | 0.31428571428571417 | 1 | 0 | 0 | 1 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 |
| 4.0 | 0.0 | 0.56756757 | 0.5 | 0.2703862660944206 | 0.8431372549019608 | 1.0 | 0.4 | 0.896551724137931 | 0.25742574257425743 | 0.41935483870967744 | 0.1532258064516129 | 0.2857142857142857 | 1 | 0 | 1 | 0 | 1 | 0 | 0 | 1 | 1 | 0 | 0 | 1 | 1 | 0 | 1 | 0 | 0 | 1 | 1 | 0 | 1 | 0 |
| 3.0 | 4.0 | 0.85135135 | 0.25 | 0.8326180257510729 | 0.5032679738562091 | 0.28395061728395066 | 0.3333333333333333 | 0.37931034482758624 | 0.47524752475247517 | 0.45161290322580644 | 0.4274193548387097 | 0.31428571428571417 | 0 | 1 | 1 | 0 | 1 | 0 | 1 | 0 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 1 | 0 | 0 | 1 | 1 | 0 |
| 3.0 | 2.0 | 0.90540541 | 1.0 | 0.9656652360515021 | 0.5228758169934641 | 0.6419753086419753 | 0.6666666666666666 | 0.0 | 0.1485148514851484 | 0.22580645161290322 | 0.21774193548387097 | 0.0571428571428572 | 1 | 0 | 1 | 0 | 0 | 1 | 1 | 0 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 |
| 4.0 | 0.0 | 0.56756757 | 0.5 | 0.1072961373390558 | 1.0 | 0.9012345679012346 | 0.5333333333333333 | 0.31034482758620685 | 0.20792079207920794 | 0.2903225806451613 | 0.20967741935483872 | 0.11428571428571425 | 0 | 1 | 1 | 0 | 1 | 0 | 1 | 0 | 0 | 1 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 |
| 4.0 | 1.0 | 0.67567568 | 0.25 | 0.6008583690987125 | 0.10457516339869281 | 0.16049382716049382 | 0.5333333333333333 | 0.31034482758620685 | 0.831

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

The model hallucinated and only returned up to row 10 of the input I gave it in markdown format. A way to mitigate this would be to chunk the data into smaller portions and have the model convert the smaller chunks and then manually combine them. You could also prompt the model with more specific constraints/formatting for the data it should be outputting.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [16]:
# Code to rename all the columns in the dataset

name_mapping = {
    "al": "albumin", 
    "su": "sugar", 
    "age": "age", 
    "bp": "blood_pressure", 
    "bgr": "blood_glucose_random", 
    "bu": "blood_urea", 
    "sc": "serum_creatinine", 
    "sod": "sodium", 
    "pot": "potassium", 
    "hemo": "hemoglobin", 
    "pcv": "packed_cell_volume", 
    "wbcc": "white_blood_cell_count", 
    "rbcc": "red_blood_cell_count", 
    "rbc_abnormal": "red_blood_cells_abnormal", 
    "rbc_normal": "red_blood_cells_normal", 
    "pc_abnormal": "pus_cell_abnormal", 
    "pc_normal": "pus_cell_normal", 
    "pcc_notpresent": "pus_cell_clumps_notpresent", 
    "pcc_present": "pus_cell_clumps_present", 
    "ba_notpresent": "bacteria_notpresent", 
    "ba_present": "bacteria_present", 
    "htn_no": "hypertension_no", 
    "htn_yes": "hypertension_yes", 
    "dm_no": "diabetes_mellitus_no", 
    "dm_yes": "diabetes_mellitus_yes", 
    "cad_no": "coronary_artery_disease_no", 
    "cad_yes": "coronary_artery_yes", 
    "appet_good": "appetite_good", 
    "appet_poor": "appetite_poor", 
    "pe_no": "pedal_edema_no", 
    "pe_yes": "pedal_edema_yes", 
    "ane_no": "anemia_no", 
    "ane_yes": "anemia_yes", 
    "Target_ckd": "Target_chronic_kidney_disease", 
    "Target_notckd": "Target_not_chronic_kidney_disease"
}

data.rename(columns=name_mapping, inplace=True)

print(data)

     albumin  sugar       age  blood_pressure  blood_glucose_random  \
0        2.0    0.0  0.743243            0.50              0.442060   
1        2.0    2.0  0.770270            1.00              0.901288   
2        2.0    0.0  0.905405            0.50              0.785408   
3        4.0    0.0  0.202703            0.75              0.158798   
4        4.0    2.0  0.783784            1.00              0.399142   
..       ...    ...       ...             ...                   ...   
128      0.0    0.0  0.189189            0.25              0.227468   
129      0.0    0.0  0.216216            0.00              0.115880   
130      0.0    0.0  0.513514            0.00              0.107296   
131      0.0    0.0  0.702703            0.50              0.128755   
132      0.0    0.0  0.689189            0.00              0.150215   

     blood_urea  serum_creatinine    sodium  potassium  hemoglobin  ...  \
0      0.901961          0.432099  0.500000   0.793103    0.000000  ... 

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [17]:
# Converted SQL to Pandas code

result = data.groupby('Target_chronic_kidney_disease').size().reset_index(name='count')
print(result)

   Target_chronic_kidney_disease  count
0                              0    111
1                              1     22


*Describe what the above code does here*

The above code counts the number of entries for each Target (chronic kidney disease and not chronic kidney disease). 111 entries in the table are associated with not chronic kidney disease, and 22 entires are associated with chronic kidney disease.